### ISSUE TITLE
Loop unrolling during backpropagation

### ISSUE THREAD

**Issue by kadir-gunel** [denizyuret/Knet.jl#200]
Trying to implement RNN by my self. What I did so far is : 

```julia

path = pwd()*"/utils.jl"
include(path)

module VANILLA

using Knet
using AutoGrad: getval
using Knet: sigm_dot, tanh_dot

onehots(atype, dim) = atype{Float32, 2}(eye(dim))
weight(atype, dims...) = atype{Float32, 2}(xavier(dims...))
bias(atype, dims...)= atype{Float32, 2}(ones(dims...))


H = 100
seq_length = 20
atype = KnetArray

function initModel(atype, H, D)
    model = []
    push!(model, weight(atype, H, D)) # Wxh
    push!(model, weight(atype, H, H)) # Whh
    push!(model, weight(atype, D, H)) # Why
    push!(model, bias(atype, H, 1)) # bh
    push!(model, bias(atype, D, 1)) # by
    return model
end


function rnn(input, model, hprev)
    return tanh_dot(model[1] * input + model[2] * hprev + model[4])
end

function loss(sentence, model, we)
    hprev = atype{Float32}(zeros(H, 1))
    totloss = 0
    # for (input, pred) in zip(sentence[1:end-1], sentence[2:end])
    hprev = rnn(hcat(we[1:end, input]), model, hprev)
    yhat = model[3] * hprev + model[5]
    ynorm = logp(yhat, 1)
    totloss += -sum(we[1:end, pred] .* ynorm) / size(ynorm, 2)
    #end
    return totloss
end

rollback_loss = grad(loss)

file = "/home/kadir/Desktop/deneme.txt"
# pre-processing
lookup, ilookup = utils.createLookUp(file)
s2i = utils.toInteger(file, lookup)
D = length(lookup)
model = initModel(atype, H, D)
we = onehots(atype, D)
optimizer = Vector{Adam}(length(model))
map((i) -> optimizer[i] = Adam(lr=0.1), range(1,length(model)))
errSent = Vector{Int64}(length(s2i))
grads = rollback_loss(s2i[1], model, we)
update!(model, grads, optimizer)
loss(s2i[1], model, we)
end
```
And when I comment out the for loop inside loss function I get `` no method matching getindex(::Knet.KnetArray{Float32,2}, ::UnitRange{Int64}, ::AutoGrad.Rec{Int64} `` error. 

Is there any way to make it work inside a for loop ? 
B.R.

p.s. Sorry about changing the title multiple times.
